File: week6/Chapter_8_Refactoring/csharp/src/OopOrganization.Chapter8/Refactoring/Policies.cs
namespace OopOrganization.Refactoring;

public sealed record Customer(string Id, string Name, string State);

public sealed class DiscountPolicy
{
    public Money Discount(Money subtotal)
        => subtotal.Amount > 100m ? subtotal.Mul(0.05m) : new Money(0m, subtotal.Currency);
}

public sealed class ShippingPolicy
{
    public Money Shipping(IEnumerable<ILineItem> items, Money subtotal)
    {
        var hasShippable = items.Any(i => i.IsShippable);
        if (!hasShippable) return new Money(0m, subtotal.Currency);
        return subtotal.Amount > 75m ? new Money(0m, subtotal.Currency) : new Money(9.95m, subtotal.Currency);
    }
}

public sealed class TaxPolicy
{
    public Money Tax(Customer customer, Money taxable)
        => customer.State == "PA" ? taxable.Mul(0.06m) : new Money(0m, taxable.Currency);
}

File: week6/Chapter_8_Refactoring/javascript/src/afterRefactor/policies.js
import { Money } from "./money.js";

export class Customer {
  constructor(id, name, state) { this.id = id; this.name = name; this.state = state; }
}

export class DiscountPolicy {
  discount(subtotal) { return subtotal.amount > 100 ? subtotal.mul(0.05) : new Money(0, subtotal.currency); }
}

export class ShippingPolicy {
  shipping(items, subtotal) {
    const hasShippable = items.some((i) => i.isShippable);
    if (!hasShippable) return new Money(0, subtotal.currency);
    return subtotal.amount > 75 ? new Money(0, subtotal.currency) : new Money(9.95, subtotal.currency);
  }
}

export class TaxPolicy {
  tax(customer, taxable) { return customer.state === "PA" ? taxable.mul(0.06) : new Money(0, taxable.currency); }
}
